### PR TITLE
[compiler-v2] Enable empty field list

### PR DESCRIPTION
**Pull request by wrwg**
## Description

Enables struct declarations of the form `struct T;`, which is interpreted as `struct T();`. This makes the declaration of type tags more concise.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Added new tests

### CODE REVIEW

**Comment by wrwg** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14415?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14415** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14415?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14411** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14411?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#14414](https://github.com/aptos-labs/aptos-core/pull/14414) <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14414?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @wrwg and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: third_party/move/move-compiler-v2/tests/checking/typing/struct_no_field_list.move [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module 0x42::m {
+
+    struct S has copy, drop;
+
+    fun f(_s: S): S {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/typing/struct_no_field_list.move:5]: Done

File: third_party/move/move-compiler-v2/tests/checking/positional_fields/decl_ok.exp [hunk 2]
```diff
@@ -9,7 +9,6 @@ module 0x42::test {
         }
     }
     struct S1 {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/positional_fields/decl_ok.exp]: Yeah, as we discussed in person, this is actually not working, so undone this part of the change.

**Review by vineethk — APPROVED** [review]

**Comment by vineethk** [comment]
@wrwg Might want to change title for other reviewers.

**Review by fEst1ck — APPROVED** [review]

**Review by fEst1ck — APPROVED** [review]

File: third_party/move/move-compiler-v2/tests/checking/typing/struct_no_field_list.move [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+        // dropping paramter, constructing new value
+        S
+    }
+    
+    fun d() {
+        let S = S;
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/checking/typing/struct_no_field_list.move:11]: Do we really want to enable these?  They do nothing, so are likely to be some sort of user error if they appear.

**Review by brmataptos — COMMENTED** [review]
I don't think enabling this kind of thing is going to help users.  Can you explain more why you like it?

> Enables struct declarations of the form struct T;, which is interpreted as struct T();. This makes the declaration of type tags more concise.

Can you provide an example of such a usage?

File: third_party/move/move-compiler-v2/tests/checking/typing/struct_no_field_list.move (hunk 3 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/typing/struct_no_field_list.move:11]: I'm not sure what you mean with enable. This possible from the logic of the language, whether it makes sense to write something like this is another question. Rust certainly allows this.

**Pull request merged**